backend.go

```diff
@@ -123,70 +123,83 @@ func (b *backend) resetClient() {
 	b.kmsClientCreateTime = time.Unix(0, 0).UTC()
 }
 
+// validClient returns a boolean indicating if the attached client is still
+// valid. This is not safe to be called concurrently. Callers must allocate an
+// RLock on kmsClientLock before accessing.
+func (b *backend) validClient() bool {
+	return b.kmsClient != nil && time.Now().UTC().Sub(b.kmsClientCreateTime) < b.kmsClientLifetime
+}
+
 // KMSClient creates a new client for talking to the GCP KMS service.
 func (b *backend) KMSClient(s logical.Storage) (*kmsapi.KeyManagementClient, func(), error) {
-	// If the client already exists and is valid, return it
-	b.kmsClientLock.RLock()
-	if b.kmsClient != nil && time.Now().UTC().Sub(b.kmsClientCreateTime) < b.kmsClientLifetime {
-		closer := func() { b.kmsClientLock.RUnlock() }
-		return b.kmsClient, closer, nil
-	}
-	b.kmsClientLock.RUnlock()
-
-	// Acquire a full lock. Since all invocations acquire a read lock and defer
-	// the release of that lock, this will block until all clients are no longer
-	// in use. At that point, we can acquire a globally exclusive lock to close
-	// any connections and create a new client.
-	b.kmsClientLock.Lock()
-
-	b.Logger().Debug("creating new KMS client")
-
-	// Attempt to close an existing client if we have one.
-	b.resetClient()
-
-	// Get the config
-	config, err := b.Config(b.ctx, s)
-	if err != nil {
-		b.kmsClientLock.Unlock()
-		return nil, nil, err
-	}
-
-	// If credentials were provided, use those. Otherwise fall back to the
-	// default application credentials.
-	var creds *google.Credentials
-	if config.Credentials != "" {
-		creds, err = google.CredentialsFromJSON(b.ctx, []byte(config.Credentials), config.Scopes...)
-		if err != nil {
-			b.kmsClientLock.Unlock()
-			return nil, nil, errwrap.Wrapf("failed to parse credentials: {{err}}", err)
+	for {
+		// If the client already exists and is valid, return it
+		b.kmsClientLock.RLock()
+		if b.validClient() {
+			closer := func() { b.kmsClientLock.RUnlock() }
```

Comment:
I haven't looked at all the places this client is used, but i'm curious if it is required to return the closer function. Since we are returning a pointer any new clients will be a distinct object, and only new requests will get the newly created client. Unless the client in mutable somewhere i don't think this should be required. Removing this would also remove the need for the for loop above since we could just return the client while holding the write lock.

Reply:
The client is mutable, yea